### PR TITLE
Make FE_EVAL_FACTORY_DEGREE_MAX more robust

### LINKED ISSUE #14106: Robustness of FEEvaluation's pre-compiled evaluation routines
In https://github.com/exadg/exadg/issues/241, we discussed the following problem:
- The ExaDG code wants more polynomial degrees inside the FEEvaluation evaluators to be pre-compiled with high efficiency (polynomial degree and number of quadrature points are compile-time constants), so it compiles deal.II's file https://github.com/dealii/dealii/blob/master/include/deal.II/matrix_free/evaluation_template_factory.templates.h with a different setting of https://github.com/dealii/dealii/blob/dfde63a25cf241f8ef3e67f8e247258cd84de634/include/deal.II/matrix_free/evaluation_template_factory_internal.h#L23-L25
- The duplicate symbols within `libdeal_II.so` and `libexadg.so` are resolved differently between different code components. This resulted in run time errors, because the check https://github.com/dealii/dealii/blob/cfebfb5e1114f93755136675369d0c6b89829304/include/deal.II/matrix_free/fe_evaluation.h#L8643-L8645 which goes to https://github.com/dealii/dealii/blob/dfde63a25cf241f8ef3e67f8e247258cd84de634/include/deal.II/matrix_free/evaluation_template_factory.templates.h#L32-L40 went to code in `libexadg.so`, whereas the code in the same function just a few lines below, https://github.com/dealii/dealii/blob/cfebfb5e1114f93755136675369d0c6b89829304/include/deal.II/matrix_free/fe_evaluation.h#L8673-L8682 called into the code of `libdeal_II.so`. This results in contradictions between the assumptions made at these two places, creating run-time errors.
- Interestingly, this worked well up to a few months ago. In the ExaDG investigation, nobody made a bisection for the cause, but I am pretty sure that the cause is #13591, where the code in the two said functions was split into different compilation units, whereas it was in the same compilation unit before.
- I know too little about how to ensure priorities in terms of the library selection when two shared libraries, `libdeal_II.so` and `libexadg.so`, provide the same symbols (as we really need to hook into a function name we know in the deal.II headers), except for `LD_PRELOAD`. However, we still recommend this in our documentation https://github.com/dealii/dealii/blob/dfde63a25cf241f8ef3e67f8e247258cd84de634/include/deal.II/matrix_free/fe_evaluation.h#L1703-L1715 

This issue is here to discuss solutions, also given that my knowledge on that end is rather limited. I think that it is good to have some polynomial degrees compiled in deal.II already as we make use of them in various places, and probably we get more. At the same time, we would like to enable the user to compile more degrees or combinations because that can be critical for an application (it is for ExaDG, for instance). So how to proceed:
- Unless we find someone with knowledge how to ensure the right selection, we might need to expose the maximal degree that gets compiled at the configure stage of deal.II. We do this already through the respective define flag https://github.com/dealii/dealii/blob/dfde63a25cf241f8ef3e67f8e247258cd84de634/.github/workflows/windows.yml#L35 but pull through a setting as a define flag to the compile line seems a bit of a hack. Do we want to enable this via a configure option?
- I think we might still allow a user code to compile something separate in the evaluation template factory, but we should try hard to not expose any inconsistency. Referring to the code in the beginning of this post, I believe we should make sure that the decision whether some setting is supported in the pre-compiled degree happen in the same function call as where we rely on this information. This won't change the risks a user takes when mixing the same symbols we provide from `libdeal_II.so`, but at least we do so consistently.

### PR DESCRIPTION
**Pull request by kronbichler**
Fixes #14106.

This implements the solution discussed in that PR: On the one hand, we make sure to always use the check `fast_evaluation_support()` compiled in the same compilation unit as the actual code we use from the factory. Furthermore, we recommend setting the define flag in the configuration of deal.II, which is robust because it does not duplicate the symbols.

### CODE REVIEW

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 1]
```diff
@@ -1709,8 +1709,16 @@ class FEEvaluationAccess<1, 1, Number, is_face, VectorizedArrayType>
...
- * evaluation/integration for a given degree/n_quadrature_points pair by calling
+ * a possibly larger set of degrees. This can both be set when configuring
+ * deal.II by passing the flag `-D FE_EVAL_FACTORY_DEGREE_MAX=8` (in case you
+ * want to compile all degrees up to eight; recommended setting) or by
+ * compiling `evaluation_template_factory.templates.h` and
+ * `evaluation_template_face_factory.templates.h` with the
+ * `FE_EVAL_FACTORY_DEGREE_MAX` overriden to the desired value. In the second
+ * option, symbols will be available twice, and it depends on your linker and
+ * dynamic library loader whether the user-specified setting takes
+ * precendence. You can check if fast evaluation/integration for a given
```
Comment (masterleinad) [inline include/deal.II/matrix_free/fe_evaluation.h:1720]: I think this is still pretty brittle and I'm not sure we really want to encourage people to do that. Maybe, mention `LD_PRELOAD` at least?

Comment (kronbichler) [reply include/deal.II/matrix_free/fe_evaluation.h:1720]: Yes, `LD_PRELOAD` is a good idea.

Comment (kronbichler) [reply include/deal.II/matrix_free/fe_evaluation.h]: I can also remove the second recommendation altogether, if we prefer that. I recommend the other option, which I wanted to achieve.

**Review by masterleinad — APPROVED** [review]
Fine with me.

**Pull request merged**